src/physics/base/PhysicsStepUtils.i.hh

```diff
@@ -124,25 +127,59 @@ CELER_FUNCTION real_type calc_energy_loss(const ParticleTrackView& particle,
  *   applicable model ID.
  */
 template<class Engine>
-CELER_FUNCTION ModelId select_model(const ParticleTrackView& particle,
-                                    const PhysicsTrackView&  physics,
-                                    Engine&                  rng)
+CELER_FUNCTION ProcessIdModelId
+select_process_and_model(const ParticleTrackView& particle,
+                         const PhysicsTrackView&  physics,
+                         Engine&                  rng)
 {
-    if (physics.interaction_mfp() > 0)
-    {
-        // Nonzero MFP to interaction -- no interaction model
-        return {};
-    }
+    // Nonzero MFP to interaction -- no interaction model
+    CELER_EXPECT(physics.interaction_mfp() <= 0);
 
     // Sample ParticleProcessId from physics.per_process_xs()
-    (void)sizeof(rng);
 
-    // Get ModelGroup from physics.models(ppid);
+    auto      total_macro_xs = physics.macro_xs();
+    real_type prob           = generate_canonical(rng) * total_macro_xs;
+    real_type accum          = 0.0;
 
-    // Find ModelId corresponding to energy bin
-    (void)sizeof(particle.energy());
+    for (auto ppid : range(ParticleProcessId{physics.num_particle_processes()}))
+    {
+        accum += physics.per_process_xs(ppid);
+        if (accum >= prob)
+        {
+            // Select the model and return;
+            auto find_model = physics.make_model_finder(ppid);
+
+            // Does the energy change between the time the per_process_xs was
+            // calculated and now?  Does the energy enters in the calculation
+            // of the cross-section? What happens to the cross section per
+            // process if there no model covering that energy range?
+            // See https://github.com/celeritas-project/celeritas/pull/165#issuecomment-790214691
+            // for some discussion, including this:
+            //
+            // Look up the model for the selected process at the reduced
+            // energy E'. If there's not one -- i.e. the along-step energy
+            // loss causes us to go below the lower threshold for the selected
+            // process -- we could simply zero out the cross section for that
+            // process and jump back up to the step above (resampling
+            // available processes). If we did the cutoff step correctly then
+            // there should be at least one process with a valid model.
+            //
+            // but
+            //
+            // Some of the glue pieces (cutoff testing, at rest) aren't yet in
+            // place, and I think for the moment we can skip the "resample if
+            // below process energy threshold" since I don't think there are
+            // any processes that have thresholds (aside from gamma, which has
+            // no continuous energy loss processes).
+
```

Comment:
Can you move this lengthy comment to the function documentation above line 130 (and leave a one-line comment here referencing the function docs)? A large block of text in the middle of the function makes it harder to read.